src/matchbox/server/postgresql/utils/query.py

```diff
@@ -364,3 +401,271 @@ def query(
+            Contains,
+            unnested_clusters,
+            unnested_clusters.c.hash == Contains.child,
+            isouter=True,
+        )
+        .where(Contains.parent == highest_parent)
+    )
+
+    # Add valid clusters filter if provided
+    if valid_clusters is not None:
+        base = base.where(Contains.child.in_(select(valid_clusters.c.cluster)))
+
+    hierarchy_down = base.cte("hierarchy_down", recursive=True)
+
+    # Recursive case: Get both intermediate nodes AND their leaf records
+    recursive = (
+        select(
+            hierarchy_down.c.parent,
+            Contains.child.label("child"),
+            (hierarchy_down.c.level + 1).label("level"),
+            unnested_clusters.c.dataset.label("dataset"),
+            unnested_clusters.c.id.label("id"),
+        )
+        .select_from(hierarchy_down)
+        .join_from(
+            hierarchy_down,
+            Contains,
+            Contains.parent == hierarchy_down.c.child,
+        )
+        .join_from(
+            Contains,
+            unnested_clusters,
+            unnested_clusters.c.hash == Contains.child,
+            isouter=True,
+        )
+        .where(hierarchy_down.c.id.is_(None))  # Only recurse on non-leaf nodes
+    )
+
+    # Add valid clusters filter to recursive part if provided
+    if valid_clusters is not None:
+        recursive = recursive.where(
+            Contains.child.in_(select(valid_clusters.c.cluster))
+        )
+
+    return hierarchy_down.union_all(recursive)
+
+
+def match(
+    source_id: str,
+    source: str,
+    target: str | list[str],
+    resolution: str,
+    engine: Engine,
+    threshold: float | dict[str, float] | None = None,
+) -> Match | list[Match]:
+    """Matches an ID in a source dataset and returns the keys in the targets.
+
+    To accomplish this, the function:
+
+    * Reconstructs the resolution lineage from the specified resolution
+    * Iterates through each target, and
+        * Retrieves its cluster hash according to the resolution
+        * Retrieves all other IDs in the cluster in the source dataset
+        * Retrieves all other IDs in the cluster in the target dataset
+    * Returns the results as Match objects, one per target
+    """
+    # Split source and target into schema/table
+    targets = [target] if isinstance(target, str) else target
+    target_pairs = [get_schema_table_names(t, validate=True) for t in targets]
+
+    with Session(engine) as session:
+        # Get source, target and truth resolutions
+        source_resolution = source_to_dataset_resolution(source, session)
+        target_resolutions = []
+        for target in targets:
+            target_resolutions.append(source_to_dataset_resolution(target, session))
+        truth_resolution = (
+            session.query(Resolutions).filter(Resolutions.name == resolution).first()
+        )
+        if truth_resolution is None:
+            raise MatchboxResolutionError(f"Resolution {resolution} not found")
```

Comment:
that's the same as the default error message, just pass resolution as an argument